Waku handshake RLP key variable initialisation

Samyoul opened the pull request:
## Why make the change?

This PR is the first attempt at resolving the issue reported here https://github.com/status-im/status-go/issues/1929, however this PR has been superseded by https://github.com/status-im/status-go/pull/1938.

## What's changed?

This PR is now is essentially a refactor of variable initialisation and the declaration of a new type. 

### Refactor

Variable declaration

- `waku.idxFieldKey map[int]statusOptionKey{}`
- `waku.keyFieldIdx map[statusOptionKey]int{}`

### Added

Type
- `waku.statusOptionKey`

Functions
- `waku.initRLPKeyFields()`
- `waku.init()`

Tests
- `TestInitRLPKeyFields()`

This test explicitly checks that the variable initialisation matches the expected parse result of the `statusOptions` struct.

Samyoul committed:
Added IDE file to gitignore

Samyoul committed:
Preliminary fix for GH-1929

ghost commented:
Hey @Samyoul, and thank you so much for making your first pull request in status-go! :heart: Please help us make your experience better by filling out this brief questionnaire https://goo.gl/forms/uWqNcVpVz7OIopXg2

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (24)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | f110280c | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/1/) | 2020-04-07 12:51:45 | ~5 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200407-124633-f11028.zip) |
| :heavy_check_mark: | f110280c | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/1/) | 2020-04-07 12:53:45 | ~7 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200407-124625-f11028.aar) |
| :heavy_check_mark: | f110280c | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/1/) | 2020-04-07 12:54:26 | ~7 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200407-124643-f11028.zip) |
| | | | | | | |
| :heavy_check_mark: | 6f557a02 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/2/) | 2020-04-07 13:20:44 | ~53 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200407-131950-6f557a.zip) |
| :heavy_check_mark: | 6f557a02 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/2/) | 2020-04-07 13:22:49 | ~3 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200407-131936-6f557a.zip) |
| :heavy_check_mark: | 6f557a02 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/2/) | 2020-04-07 13:25:57 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200407-131922-6f557a.aar) |
| | | | | | | |
| :heavy_check_mark: | e4ce0c59 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/3/) | 2020-04-07 17:26:18 | ~46 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200407-172531-e4ce0c.zip) |
| :heavy_check_mark: | e4ce0c59 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/3/) | 2020-04-07 17:27:53 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200407-172522-e4ce0c.zip) |
| :heavy_check_mark: | e4ce0c59 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/3/) | 2020-04-07 17:31:08 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200407-172513-e4ce0c.aar) |
| | | | | | | |
| :heavy_check_mark: | 3fa9145a | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/4/) | 2020-04-08 22:48:48 | ~42 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200408-224805-3fa914.zip) |
| :heavy_check_mark: | 3fa9145a | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/4/) | 2020-04-08 22:50:24 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200408-224756-3fa914.zip) |
| :heavy_check_mark: | 3fa9145a | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/4/) | 2020-04-08 22:53:36 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200408-224748-3fa914.aar) |
| | | | | | | |
| :heavy_check_mark: | bbe97199 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/5/) | 2020-04-09 11:23:13 | ~50 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200409-112223-bbe971.zip) |
| :heavy_check_mark: | bbe97199 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/5/) | 2020-04-09 11:25:44 | ~3 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200409-112215-bbe971.zip) |
| :heavy_check_mark: | bbe97199 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/5/) | 2020-04-09 11:28:02 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200409-112205-bbe971.aar) |
| | | | | | | |
| :heavy_check_mark: | 01e1db13 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/6/) | 2020-04-09 15:27:45 | ~51 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200409-152653-01e1db.zip) |
| :heavy_check_mark: | 01e1db13 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/6/) | 2020-04-09 15:29:18 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200409-152644-01e1db.zip) |
| :heavy_check_mark: | 01e1db13 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/6/) | 2020-04-09 15:36:16 | ~9 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200409-152636-01e1db.aar) |
| | | | | | | |
| :heavy_check_mark: | ad1fe07f | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/7/) | 2020-04-09 22:42:46 | ~47 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200409-224158-ad1fe0.zip) |
| :heavy_check_mark: | ad1fe07f | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/7/) | 2020-04-09 22:44:18 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200409-224149-ad1fe0.zip) |
| :heavy_check_mark: | ad1fe07f | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/7/) | 2020-04-09 22:47:37 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200409-224141-ad1fe0.aar) |
| | | | | | | |
| :heavy_check_mark: | 1f140d43 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/8/) | 2020-04-09 22:49:08 | ~36 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200409-224831-1f140d.zip) |
| :heavy_check_mark: | 1f140d43 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/8/) | 2020-04-09 22:50:46 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200409-224822-1f140d.zip) |
| :heavy_check_mark: | 1f140d43 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/8/) | 2020-04-09 22:53:52 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200409-224813-1f140d.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 9c156562 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/9/) | 2020-04-14 21:54:26 | ~55 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200414-215330-9c1565.zip) |
| :heavy_check_mark: | 9c156562 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/9/) | 2020-04-14 21:56:02 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200414-215322-9c1565.zip) |
| :heavy_check_mark: | 9c156562 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/9/) | 2020-04-14 21:59:07 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200414-215314-9c1565.aar) |
| | | | | | | |
| :heavy_check_mark: | 5c26958e | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1931/10/) | 2020-04-17 11:49:31 | ~44 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200417-114846-5c2695.zip) |
| :heavy_check_mark: | 5c26958e | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1931/10/) | 2020-04-17 11:54:36 | ~5 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200417-114838-5c2695.zip) |
| :heavy_check_mark: | 5c26958e | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1931/10/) | 2020-04-17 11:54:45 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200417-114830-5c2695.aar) |

Samyoul committed:
Added handling of parseStatuOptionKeys() error

Samyoul committed:
Added new test and used static var declarations for keyFieldIdx

Samyoul committed:
Added handling for mixed RLP key data types

This the initial implementation, a number of known test fails have been introduced with this commit and will be resolved as a matter of priority.

adambabik commented on waku/handshake.go:
```diff
@@ -6,12 +6,50 @@ import (
 )
 
-var defaultMinPoW = math.Float64bits(0.001)
+type statusOptionKey uint
```
```suggestion
// statusOptionKey is a current type used in statusOptions as a key.
type statusOptionKey uint
```

adambabik commented on waku/handshake.go:
```diff
@@ -6,12 +6,50 @@ import (
-var defaultMinPoW = math.Float64bits(0.001)
+type statusOptionKey uint
+
+type statusOptionKeyType uint
```
```suggestion
// statusOptionKeyType is a type of a statusOptions key used for a particular instance of statusOptions struct.
type statusOptionKeyType uint
```

adambabik commented on waku/handshake.go:
```diff
@@ -6,12 +6,50 @@ import (
+}
+
+const (
+	sOKTS statusOptionKeyType = iota // Status Option Key Type String
```
```suggestion
	sOKTS statusOptionKeyType = iota + 1 // Status Option Key Type String
```

Avoid zero value as it's impossible to tell whether it's an uninitialized variable or actual value.

Samyoul replied:
Good point

adambabik commented on waku/handshake.go:
```diff
@@ -158,6 +206,58 @@ loop:
+	o.keyTypeMapping.keyFieldIdx[ktt.Key] = ktt
+}
+
+func (k statusOptionKeyToType) decodeStream(s *rlp.Stream) error {
```
```suggestion
func (k *statusOptionKeyToType) decodeStream(s *rlp.Stream) error {
```

Shouldn't it be a pointer receiver? You assign values to `k` below.

Samyoul replied:
Ah, yes. Thank you. I can't believe I missed this.

adambabik commented on waku/handshake.go:
```diff
@@ -6,12 +6,50 @@ import (
+
+var (
+	defaultMinPoW = math.Float64bits(0.001)
+	idxFieldKey   = map[int]statusOptionKey{
```
I gave it a though and we kinda rely on internal Go implementation here. It should not change but who knows...

I would do it this way:
1. When encoding, go through all fields with `rlp:"key=N"`.
2. When decoding, create a method `statusOptions.setField(key, value)` which would set a value for a field based on `key`, i.e. for a field with `rlp:"key={key}"`.

Samyoul replied:
>I gave it a though and we kinda rely on internal Go implementation here. It should not change but who knows...

Ok, I can implement the `init()` func we discussed earlier.

adambabik commented on waku/handshake.go:
```diff
@@ -26,6 +64,7 @@ type statusOptions struct {
 	ConfirmationsEnabled *bool       `rlp:"key=3"`
 	RateLimits           *RateLimits `rlp:"key=4"`
 	TopicInterest        []TopicType `rlp:"key=5"`
+	keyTypeMapping       keyTypeMapping
```
Based on the comment above, we would not need this. We would only need a field `keyType statusOptionKeyType`.

Samyoul replied:
Isn't this presuming that all the keys are of the same type? What if some are `string` and some are `uint`?

Samyoul replied:
Or would there be a `statusOptionKeyType` for each `statusOptionKey`?

adambabik replied:
> Isn't this presuming that all the keys are of the same type? What if some are string and some are uint?

This is not possible. The current version always sends keys as strings and this will always send keys as ints unless it received keys as strings then it will also reply with string keys.

Samyoul replied:
Ok. I will sort this.

Samyoul committed:
Apply suggestions from code review

Co-Authored-By: Adam Babik <a.babik@designfortress.com>

Samyoul committed:
Implemented RLP key field initialisation

This replaces the statically declared values, which previously replaced function varialbes see https://github.com/status-im/status-go/pull/1931#discussion_r406123786

adambabik commented on waku/handshake.go:
```diff
@@ -26,6 +53,39 @@ type statusOptions struct {
+			continue
+		}
+
+		keys := strings.Split(rlpTag, "=")
```
I would add a check:
```go
if len(keys) != 2 || keys[0] != "key" {
    panic("invalid value of \"rlp\" tag, expected \"key=N\" where N is uint")
}
```

Samyoul replied:
Yes. I had a similar check in previous versions of this functionality but for some reason it is gone... Or maybe I just imagined it. Either way I 100% agree and will implement this.

adambabik commented on waku/handshake.go:
```diff
@@ -26,6 +53,39 @@ type statusOptions struct {
+}
+
+// initFLPKeyFields initialises the values of `idxFieldKey` and `keyFieldIdx`
+func initRLPKeyFields() error {
```
Nice 👍

Samyoul replied:
@cammellos FYI, I'd like to keep this functionality in waku. I think that it is more sustainable to have variables initialised using an actual init function.

I can revert all the changes implemented by this branch and then keep all the changes related to this specific change.

cammellos replied:
sure, sounds good to me

adambabik commented on waku/handshake.go:
```diff
@@ -97,15 +135,41 @@ func (o statusOptions) EncodeRLP(w io.Writer) error {
+			continue
+		}
+
+		var val []interface{}
```
I don't quite get it down here. I would see something like:
```go
var keyValue interface{}
if o.keyType == sOKTS {
    keyValue = fmt.Sprint(key)
} else {
    keyValue = key
}

if value != nil {
    optionsList = append(optionsList, []interface{}{keyValue, value})
}
```

Or if you want to keep `statusOptionKeyToType`:
```go
ktt := statusOptionKeyToType{Key: key, Type: o.typ}
if value != nil {
    optionsList = append(optionsList, []interface{}{ktt, value}) // statusOptionKeyToType must implement `EncodeRLP`
}
```

Samyoul replied:
I'm completely removing `statusOptionKeyToType` so all of this will be gone completely and work more like you've suggested. 😄

Samyoul committed:
Refactor to waku handshake keyType identification

Presuming that all RLP keys will be of the same type. Still failing on forward and backward compatibility, so I will fidn a solution for this next

Samyoul committed:
Added error handling for waku handshake DecodeRLP

Samyoul committed:
Added better prediction of waku RLP key type and value

Samyoul commented:
https://github.com/status-im/status-go/pull/1938 makes this branch redundant. Marking this for deletion.

Waiting to see if there is any code that is reusable. Probably the init functionality.

Samyoul committed:
Reverting most of the handshake functionality

See GH-1938

Samyoul committed:
Reverting most tests related to handshake change

Samyoul commented on waku/handshake.go:
```diff
@@ -11,7 +11,14 @@ import (
 
-var defaultMinPoW = math.Float64bits(0.001)
+// statusOptionKey is a current type used in statusOptions as a key.
+type statusOptionKey string
```
I've explicitly declared the RLP options key type here. In `waku/1` this should be changed to be `uint`

Samyoul commented on waku/handshake.go:
```diff
@@ -28,6 +35,35 @@ type statusOptions struct {
+		}
+
+		// typecast key to be of statusOptionKey type
+		keyFieldIdx[statusOptionKey(keys[1])] = i
+		idxFieldKey[i] = statusOptionKey(keys[1])
```
In `waku/1` the key needs to be parsed into a `uint` before it can be used as a `statusOptionKey`

Samyoul commented on waku/handshake.go:
```diff
@@ -28,6 +35,35 @@ type statusOptions struct {
+
+		keys := strings.Split(rlpTag, "=")
+
+		if len(keys) != 2 || keys[0] != "key" {
+			panic("invalid value of \"rlp\" tag, expected \"key=N\" where N is uint")
+		}
```
@cammellos Any objection to throwing a panic here, rather than throwing an error? I think a panic needs to be given, and in my mind it can be here or in the package `init()`.

cammellos replied:
sure, panic looks fine

Samyoul commented on .gitignore:
```diff
@@ -66,6 +66,7 @@ Session.vim
 /.idea/
 /.vscode/
 /cmd/*/.ethereum/
+*.iml
```
@cammellos I've added my IDE file to the gitignore in this pull. Any objection?

cammellos replied:
no objections :)

Samyoul commented:
Hey @cammellos and @jakubgs I'm ready for this to be merged into `develop`. When you get some time could you pass your eye over this 100% essential refactor 😄.

cammellos commented:
I will quickly test this against the production app and merge if the test is successful.

cammellos commented:
tested successfully
